File: GFG/Alex Travelling/Alex Travelling.java
class Solution {
    class Pair{
        int node;
        int weight;
        Pair(int n,int w){
            node=n;
            weight=w;
        }
    }
    int minimumCost(int[][] flights, int n, int k) {
        // Your code here
        Map<Integer,ArrayList<Pair>> map=new HashMap<>();
        for(int[]i:flights){
            int u=i[0];
            int v=i[1];
            int w=i[2];
            map.putIfAbsent(u,new ArrayList<>());
            map.get(u).add(new Pair(v,w));
        }                    
        
        int[]dis=new int[n+1];
        int ans=0;
        Queue<Pair> q=new LinkedList<>();
        q.add(new Pair(k,dis[k]));
        
        while(!q.isEmpty()){
            Pair curr=q.remove();
            int node=curr.node;
            int cost=curr.weight;
            if(!map.containsKey(node))continue;
            for(Pair p:map.get(node)){
                int neb=p.node;
                int nebCost=p.weight;
                if((dis[neb]==0 || dis[neb]>cost+nebCost) && neb!=k){
                    dis[neb]=cost+nebCost;
                    q.add(new Pair(neb,dis[neb]));
                }
            }
        }
        
        for(int i=1;i<dis.length;i++){
            if(dis[i]!=0){
                ans=Math.max(ans,dis[i]);
            }else if(i!=k){
                return -1;
            }
        }
        return ans;
    }
}

File: GFG/Alex Travelling/Alex Travelling.cpp
class Solution {
   
  public:
    int minimumCost(vector<vector<int>>& flights, int n, int k) {
        // code here
        
        unordered_map<int,list<pair<int,int>> > adj;
        for(int i = 0; i < flights.size(); i++)
        {
            int u = flights[i][0];
            int v = flights[i][1];
            int w = flights[i][2];
            
            adj[u].push_back({v,w});
            
        }
        
        vector<int> dist(n+1, INT_MAX);
        
        set<pair<int,int>> s; //distNode, node
        
        dist[k] = 0;
        s.insert({0, k});
        
        while(!s.empty())
        {
            auto top = *(s.begin());
            s.erase(s.begin());
            
            int distNode = top.first;
            int node = top.second;
            
            for(auto neigh : adj[node])
            {
                if(distNode + neigh.second < dist[neigh.first])
                {
                    dist[neigh.first] = distNode + neigh.second;
                    
                    s.insert({dist[neigh.first], neigh.first});
                }
            }
        }
        
        int ans = 0;
        
        for(int i = 1; i <= n; i++)
        {
            if(dist[i] == INT_MAX) return -1;
            
            ans = max(ans, dist[i]);
              
            // cout<<dist[i]<<" ";
        }
        return ans;
    }
};
